cpmpy/transformations/decompose_global.py

```diff
@@ -0,0 +1,100 @@
+import copy
+
+from ..expressions.globalconstraints import GlobalConstraint
+from ..expressions.core import Expression, Comparison, Operator
+from ..expressions.utils import is_any_list, eval_comparison
+from ..expressions.python_builtins import all
+from .flatten_model import flatten_constraint
+
+def decompose_global(lst_of_expr, supported=set(), supported_reif=set()):
+
+    """
+        Decomposes any global constraint not supported by the solver
+        Accepts a list of flat constraints as input
+        Returns a list of flat constraints
+
+        - supported: a set of supported global constraints or global functions
+        - supported_reified: a set of supported global constraints within a reification
+
+        Unsupported global constraints are decomposed into equivalent simpler constraints.
+        Numerical global constraints are left reified even if not part of the `supported_reified` set
+            as they can be rewritten using `cpmpy.transformations.reification.reify_rewrite`...
```

Comment:
What is the advantage of this approach?

Reply:
It means not decomposed and still reified.
The idea is the following:
if `max(x,y,z)` is a supported function in the solver (that is, `max` is in `supported`) then we can rewrite constraints as follows:
```python
bv -> max(x,y,z) <= 10
= [bv -> IV0 <= 10, max(x,y,z) == IV0]
```
So for numerical global constraints, it does not matter if they are reified or not. The rewrite happens in `reify_rewrite`. I do agree it is a bit messy but doing the `reify_rewrite` here also seems weird...

There is one edge case where we cannot do the rewrite trick: for non-total functions such as element as the following is not necessarily equivalent:
```python
bv -> [x,y,z][idx] <= 10
= [bv -> IV0 <= 10, [x,y,z][idx] == IV0]
```
as `idx` might be out of bounds if `bv` is `False`. So in that case we do have to decompose...

Reply:
Also, what is a reified numerical global? Is it `count(...) > x` or `p -> count(...) > x`? Will this method decompose the latter?

Reply:
Aha, I got it now. Thanks for the explanation! Please add the above explanation to the comments - other developers and your future self will be grateful ;)